### PR TITLE
Initial attempt at a filter view for the elections page

### PR DESCRIPTION
**Pull request by symroe**
Screenshot 2019-04-01 14 42 51

The idea here is to:

1. Make it easier to navigate the elections page when there are a lot of them. For example, if we have to deal with a EU election while the locals are stuff current, and users want to only see the EU ballots, we can link to `/elections/?election_type=europarl` (or whatever we call the EU type). 

2. Show more information on that page, like lock status and candidates. I think a table (like the home page one) works better than a huge nested unordered list

3. Remove other views that show lists of ballots with a particular status – we have "ballots with lock suggestions" and "ballots with documents that need bulk adding" views at the moment – these filters remove the need for them.

The first commit is messy, in that I've included CSS in the HTML and I've not written much in the way of tests for it. I useful review at this stage would be thinking through:

1. Does this set of filters deal with all the current views we have?
2. Should we add any more filters? We have loads of extra data, but I don't want to add tonnes of filters if they have a very small set of cases.
3. Were the other views doing anything that this doesn't do?

When we've thought about that, I can work on cleaning up / binning the code.

### CODE REVIEW

**Comment by chris48s** [comment]
This is just a first pass for comments. I've not really looked at the code in too much detail or tested locally. Hopefully some of this is of use..

At the moment we've got the `SOPNReviewRequiredView` which shows us ballots with a SOPN and no lock and no lock suggestion (i.e: areas where someone needs to go and type in the SOPN). Looking at the code, I think I can reproduce this with "has sopn" + "unlocked", but it wasn't clear to me that what "unlocked" actually means is that the ballot isn't locked _and also has no lock suggestion_.

We also have the `SuggestLockReviewListView` which shows us unlocked ballots with a lock suggestion and a SOPN where the lock suggestion wasn't made by me. This _is_ a useful thing that I wouldn't want to completely lose and as FBRC points out on slack, there isn't a combination of filters that quite gets you that.

One of the main things I think we also wanted to do here was to show ballots to (or hide ballots from) some subset of users based on whether they have a `RawPeople` object associated with them or not. I don't think we've attempted to do that here. I think actually merging all this stuff into a single elections view probably makes that conceptually harder rather than easier because it becomes less acceptable to hide a ballot from a particular class of user if we're just presenting a single generic list of ballots with no context of what situation it is to be used in.

If you do want to present this as only a single elections page with filters, I know we can link in some places to a version with particular filters pre-applied, but I feel like it might be better to present something more like a simple view and advanced view. Some of the more common combinations of filters could be presented in a dropdown as "presets", but then you can expand this as an "advanced" view instead of expecting users to construct their own query for relatively well established user journeys.

I reckon the other thing we should add is the ability to filter by date (you could do all/current/future, but a proper date picker/range is the stretch goal). One of the problems we've discussed before is that once an election becomes "not current", it basically disappears forever and that would allow us to fix that.

**Comment by JoeMitchell** [comment]
I think this is prettttty cool. l can imagine people using the election type filter in the event we get more elections. The table view is an improvement too. 

It doesn't solve the masssssively long list, but I think we decided that this wasn't a problem and that people could Ctrl/Cmd-F to victory. 

It doesn't tell me anything about the depth of data... i.e. once we've done SoPNs and locked them, I still don't know where to start focussed volunteering. But that might be better solved elsewhere.

**Comment by symroe** [comment]
Screenshot 2019-04-02 12 27 29

Ok, a little improvement after some thinking about Chris's comments. 

I reckon spelling out the filters by "actions people can take" is the way to go. The "Ready for data input" filter (bad name, ideas?) is the thing to press / link to if you're at a SoPN party and are wanting to do something. The "Locking required" button can only show for people who are able to lock. Etc.

What else should we have there?

**Comment by symroe** [comment]
Right, I've done a bunch O'work on this, and I'm afraid I've smashed in some other fixes to bugs I've found that need fixing before tomorrow.

Some images:

------
## Help in the header, depending on filter
Screenshot 2019-04-03 15 16 21
------

## Inline locking / ballot view for locking suggested posts
Screenshot 2019-04-03 15 14 26
------

I need to clean up and test the code a bit (pull the CSS out of the template, for example), but I reckon feedback now on the UX aspects would be helpful.

**Comment by symroe** [comment]
This pull request **introduces 1 alert** when merging 79b5e0afd716451d301719f202d827f2cf1d9e71 into 10515f7f6254349d1be485675b45d96c649e06f8 - [view on LGTM.com](https://lgtm.com/projects/g/DemocracyClub/yournextrepresentative/rev/pr-f81ca4b82d8f0de57d8b4baacfa40d32701a7c74)

**new alerts:**

* 1 for Unused import

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by chris48s** [comment]
Right. I'm attempting to really rapidly review this. I haven't gone over the code in a lot of detail. Beyond the test failures, I think the things I've found that are important to address are:

1. If I hit `/moderation/suggest-lock/` and I'm not in the "trusted to lock" group, it will show me ballots and tell me to lock them, but then the buttons just don't work
2. If you're going to only dole out 10 records at a time, the ballot lock page doesn't really make it clear you're only seeing 10. If I review 10, that doesn't mean I've finished - it means I have to reload the page to get another allocation, but this is unclear. Maybe we need a "give me another 10" button or something?
3. I think we still have the problem that every user who is locking gets allocated the same records to lock at the same time.

There's some other stuff I've spotted which I'm going to leave comment on, but I reckon that is the big stuff..

File: ynr/apps/moderation_queue/templates/moderation_queue/includes/lock_suggestions.html [hunk 1]
```diff
@@ -1,165 +0,0 @@
-<p>Suggestions are grouped by nomination paper</p>
```
Comment (chris48s) [inline ynr/apps/moderation_queue/templates/moderation_queue/includes/lock_suggestions.html]: is this just dead code that should have been removed in #792 ?

**Comment by symroe** [comment]
1. Yep, will add a decorator for that
2. Good point, will add
3. It's ordering by `?` (SQL `RANDOM()`) so I don't think this is a problem? Not until the last 10 anyway.

File: ynr/apps/moderation_queue/templates/moderation_queue/includes/inline-lock-table-view.html [hunk 2]
```diff
@@ -0,0 +1,48 @@
+<tr>
+  <td>
+    Suggested by<br>
```
Comment (chris48s) [inline ynr/apps/moderation_queue/templates/moderation_queue/includes/inline-lock-table-view.html]: This is functional, but if feels a bit randomly sneezed onto the page. We can probably clean this up a bit at some point

Screenshot at 2019-04-04 16-03-46

**Comment by chris48s** [comment]
> It's ordering by `?` (SQL `RANDOM()`) so I don't think this is a problem? Not until the last 10 anyway.

Happy to accept that I am wrong/on crack/trying to review a really big PR really quickly, but my reading of the code was that its `.order_by("election__election_date", "election__name")`

File: ynr/apps/moderation_queue/templates/moderation_queue/includes/lock_suggestions.html (hunk 1 above)
Comment (symroe) [reply ynr/apps/moderation_queue/templates/moderation_queue/includes/lock_suggestions.html]: Nah, it's an include I've stopped using in this PR https://github.com/DemocracyClub/yournextrepresentative/pull/863/files/79b5e0afd716451d301719f202d827f2cf1d9e71#diff-1175212586b450466bc4c4eec28fd9feL24

**Comment by symroe** [comment]
> my reading of the code was that its .order_by("election__election_date", "election__name")

Ah – maybe we're talking about different things? I'm looking at: https://github.com/DemocracyClub/yournextrepresentative/pull/863/files#diff-00230e2536068708c45e7c55919fc3b5R601

File: ynr/apps/elections/views.py [hunk 3]
```diff
@@ -68,9 +70,25 @@ class ElectionListView(TemplateView):
...
+            .annotate(memberships_count=Count("membership"))
+            .order_by("election__election_date", "election__name")
         )
+
+        f = BallotFilter(self.request.GET, qs)
+
+        context["filter"] = f
+        context["queryset"] = f.qs
+        if f.data.get("review_required") == "suggestion":
+            context["queryset"] = context["queryset"][:10]
```
Comment (chris48s) [inline ynr/apps/elections/views.py]: Nope - I think I'm just going to fast and getting confused/looking at the wrong bit of code I think..

But then my follow-on question is: do we need/want this? In the `/elections` view, presumably we just want to always show all elections?

Comment (symroe) [reply ynr/apps/elections/views.py]: Oh yeah, need to remove that 👍

**Comment by symroe** [comment]
> Couple of other bits I noticed:

1. Examples please? This is an easy fix to do as and when, so happy to do them, but I expect I'm a little blind to it
2. Done 🎉 
3. Excluded based on user, meaning you only get a red dot if _there is something for you to do_. I think this is an improvement

**Comment by chris48s** [comment]
example:

`suggestion&has_sopn=&election_type=local`

Screenshot at 2019-04-04 16-41-26

`unlocked&has_sopn=&election_type=local`

Screenshot at 2019-04-04 16-41-59

File: ynr/apps/moderation_queue/templates/moderation_queue/suggestedpostlock_review.html [hunk 4]
```diff
@@ -1,36 +1,69 @@
...
-{% endif %}
 
-{% if my_suggestions %}
-  <h2>My own lock suggestions</h2>
-  <p>Someone else needs to review these:</p>
-  {% include "moderation_queue/includes/lock_suggestions.html" with post_elections=my_suggestions %}
 
+{% if ballots %}
+  <p>Suggestions are grouped by nomination paper</p>
+  <p>Showing {{ ballots.count }} of {{ total_ballots|intcomma }} ballots with suggested locks. Reload this page to get another random set.</p>.
```
Comment (chris48s) [inline ynr/apps/moderation_queue/templates/moderation_queue/suggestedpostlock_review.html]: v minor: `</p>.` is causing a random . on its own line

**Comment by symroe** [comment]
hmm – I might be confused, but that screen short seems to be internally correct – it looks like both ballots are unlocked and one has a lock suggestion.

Maybe if there's a lock suggestion it shouldn't say "unlocked"?

**Comment by chris48s** [comment]
oh yes good point - unlocked padlock icon means lock suggestion not unlocked. Ignore that one

**Comment by symroe** [comment]
This pull request **introduces 1 alert** when merging 7177b68440701d67e2b4bbeba635a86e1212c043 into 6ac05a4b262bd553212afe66bfe7c1aed69f5a3b - [view on LGTM.com](https://lgtm.com/projects/g/DemocracyClub/yournextrepresentative/rev/pr-921e88778f606e477b8cf7714172d8e245584e1b)

**new alerts:**

* 1 for Unused import

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Pull request merged**